### PR TITLE
[quant][pt2e] XNNPACKQuantizer skip inserting observers for non-float Tensors

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114999

Summary:
att

Test Plan:
python test/test_quantization.py -k test_add_mul_long

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py [hunk 1]
```diff
@@ -927,11 +953,14 @@ def _convert_scalars_to_attrs(model: torch.fx.GraphModule) -> torch.fx.GraphModu
             prefix = "_tensor_constant_"
             get_new_attr_name = get_new_attr_name_with_prefix(prefix)
             tensor_constant_name = get_new_attr_name(model)
-            model.register_buffer(tensor_constant_name, torch.tensor(float(args[i])))
+            float_tensor = torch.tensor(float(args[i]))
+            model.register_buffer(tensor_constant_name, float_tensor)
+            fake_mode = n.meta["val"].fake_mode
```
Comment (kimishpatel) [inline torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:958]: why is this change needed?

**Review by kimishpatel — CHANGES_REQUESTED** [review]
If only float tensors are observable, then this change should apply to all ops not just binary, right?

**Comment by kimishpatel** [comment]
> > If only float tensors are observable, then this change should apply to all ops not just binary, right?
> 
> technically yes, but other ops probably don't get non-float tensor inputs right now so I think we can fix on demand

I will accept for now, but I would say this should be in the observer. If you have observer that can observer non float, let them do it.

**Review by kimishpatel — APPROVED** [review]

**Comment by jerryzh168** [comment]
> I will accept for now, but I would say this should be in the observer. If you have observer that can observer non float, let them do it.

yeah this is specific to observer, in the meantime we are making a decision in the quantizer to insert observer or not, so quantizer need to know this as well, since we are using histogram observer in xnnpack quantizer which does not work for int/long tensors, we have to skip it. But if we have a different observer that works in this case, we could use that instead as well.

or are you saying that each observer can return some structured information so that all quantizer can just use? something like: observer.supported_input_dtypes() -> List[dtype] and each quantizer can check this before hand?

**Comment by kimishpatel** [comment]
> are you saying that each observer can return some structured information so that all quantizer can just use? something like: observer.supported_input_dtypes() -> List[dtype] and each quantizer can check this before hand?

Yeah sometime like that and when quantizer still configures it with incompatible type observer can just throw error.

**Comment by jerryzh168** [comment]
> > are you saying that each observer can return some structured information so that all quantizer can just use? something like: observer.supported_input_dtypes() -> List[dtype] and each quantizer can check this before hand?
> 
> Yeah sometime like that and when quantizer still configures it with incompatible type observer can just throw error.

OK that sounds good, I think we can do that

**Review by guangy10 — APPROVED** [review]
The fix looks good to me. Thanks Jerry!

**Pull request closed without merging**